packages/postgres-database/src/simcore_postgres_database/utils_folders_v2.py

```diff
@@ -0,0 +1,916 @@
+) -> "_FolderPermissions":
+    _ = description
+    return _FolderPermissions(read=r, write=w, delete=d)
+
+
+def _only_true_permissions(permissions: _FolderPermissions) -> dict:
+    return {k: v for k, v in permissions.items() if v is True}
+
+
+def _or_reduce(x: _FolderPermissions, y: _FolderPermissions) -> _FolderPermissions:
+    return _FolderPermissions(
+        {
+            "read": x["read"] or y["read"],
+            "write": x["write"] or y["write"],
+            "delete": x["delete"] or y["delete"],
+        }
+    )
+
+
+def _or_dicts_list(dicts: Iterable[_FolderPermissions]) -> _FolderPermissions:
+    if not dicts:
+        return _make_permissions()
+    return reduce(_or_reduce, dicts)
+
+
+class _BasePermissions:
+    LIST_FOLDERS: ClassVar[_FolderPermissions] = _make_permissions(r=True)
+    # NOTE `LIST_PROJECTS` bypasses these access rights and always lists all projects in the fodler
+
+    CREATE_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(w=True)
+    ADD_PROJECT_TO_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(w=True)
+
+    SHARE_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+    UPDATE_FODLER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+    DELETE_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+    REMOVE_PROJECT_FROM_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+
+    _MOVE_FOLDER_SOURCE: ClassVar[_FolderPermissions] = _make_permissions(
+        d=True,
+        description="apply to folder form which data is copied",
+    )
+    _MOVE_FOLDER_DESTINATION: ClassVar[_FolderPermissions] = _make_permissions(
+        w=True, description="apply to folder to which data will be copied"
+    )
+    MOVE_FOLDER: ClassVar[_FolderPermissions] = _or_dicts_list(
+        [_MOVE_FOLDER_SOURCE, _MOVE_FOLDER_DESTINATION]
+    )
+
+
+NO_ACCESS_PERMISSIONS: _FolderPermissions = _make_permissions()
+
+VIEWER_PERMISSIONS: _FolderPermissions = _or_dicts_list(
+    [
+        _BasePermissions.LIST_FOLDERS,
+    ]
+)
+EDITOR_PERMISSIONS: _FolderPermissions = _or_dicts_list(
+    [
+        VIEWER_PERMISSIONS,
+        _BasePermissions.CREATE_FOLDER,
+        _BasePermissions.ADD_PROJECT_TO_FOLDER,
+    ]
+)
+OWNER_PERMISSIONS: _FolderPermissions = _or_dicts_list(
+    [
+        EDITOR_PERMISSIONS,
+        _BasePermissions.SHARE_FOLDER,
+        _BasePermissions.UPDATE_FODLER,
+        _BasePermissions.DELETE_FOLDER,
+        _BasePermissions.REMOVE_PROJECT_FROM_FOLDER,
+        _BasePermissions.MOVE_FOLDER,
+    ]
+)
+
+_ROLE_TO_PERMISSIONS: dict[FolderAccessRole, _FolderPermissions] = {
+    FolderAccessRole.NO_ACCESS: NO_ACCESS_PERMISSIONS,
+    FolderAccessRole.VIEWER: VIEWER_PERMISSIONS,
+    FolderAccessRole.EDITOR: EDITOR_PERMISSIONS,
+    FolderAccessRole.OWNER: OWNER_PERMISSIONS,
+}
+
+
+def _hash_permissions(permissions: _FolderPermissions) -> tuple:
+    return tuple(permissions.items())
```

Comment:
if _FolderPermissions was a frozen dataclass you would not need this hashing at all -> simpler code